external/float32/float32_stubs.c

```diff
@@ -0,0 +1,332 @@
+}
+
+int32_t compiler_float32_abs(int32_t i)
+{
+    return int32_of_float32(fabsf(float32_of_int32(i)));
+}
+
+int32_t compiler_float32_add(int32_t i, int32_t j)
+{
+    return int32_of_float32(float32_of_int32(i) + float32_of_int32(j));
+}
+
+int32_t compiler_float32_sub(int32_t i, int32_t j)
+{
+    return int32_of_float32(float32_of_int32(i) - float32_of_int32(j));
+}
+
+int32_t compiler_float32_mul(int32_t i, int32_t j)
+{
+    return int32_of_float32(float32_of_int32(i) * float32_of_int32(j));
+}
+
+int32_t compiler_float32_div(int32_t i, int32_t j)
+{
+    return int32_of_float32(float32_of_int32(i) / float32_of_int32(j));
+}
+
+int32_t compiler_float32_mod(int32_t i, int32_t j)
+{
+    return int32_of_float32(fmodf(float32_of_int32(i), float32_of_int32(j)));
+}
+
+value compiler_float32_compare(int32_t i, int32_t j)
+{
+    /* See caml_float_compare_unboxed */
+    float f = float32_of_int32(i);
+    float g = float32_of_int32(j);
+    intnat res =
+        (intnat)(f > g) - (intnat)(f < g) + (intnat)(f == f) - (intnat)(g == g);
+    return Val_int(res);
+}
+
+value compiler_float32_equal(int32_t i, int32_t j)
+{
+    return Val_bool(float32_of_int32(i) == float32_of_int32(j));
+}
+
+int32_t compiler_float32_of_float(double d)
+{
+    return int32_of_float32((float)d);
+}
+
+double compiler_float32_to_float(int32_t i)
+{
+    return (double)float32_of_int32(i);
+}
+
+value compiler_float32_neg_boxed(value i)
+{
+    return caml_copy_int32(compiler_float32_neg(Int32_val(i)));
+}
+
+value compiler_float32_abs_boxed(value i)
+{
+    return caml_copy_int32(compiler_float32_abs(Int32_val(i)));
+}
+
+value compiler_float32_add_boxed(value i, value j)
+{
+    return caml_copy_int32(compiler_float32_add(Int32_val(i), Int32_val(j)));
+}
+
+value compiler_float32_sub_boxed(value i, value j)
+{
+    return caml_copy_int32(compiler_float32_sub(Int32_val(i), Int32_val(j)));
+}
+
+value compiler_float32_mul_boxed(value i, value j)
+{
+    return caml_copy_int32(compiler_float32_mul(Int32_val(i), Int32_val(j)));
+}
+
+value compiler_float32_div_boxed(value i, value j)
+{
+    return caml_copy_int32(compiler_float32_div(Int32_val(i), Int32_val(j)));
+}
+
+value compiler_float32_mod_boxed(value i, value j)
+{
+    return caml_copy_int32(compiler_float32_mod(Int32_val(i), Int32_val(j)));
+}
+
+value compiler_float32_compare_boxed(value i, value j)
+{
+    return compiler_float32_compare(Int32_val(i), Int32_val(j));
+}
+
+value compiler_float32_equal_boxed(value i, value j)
+{
+    return compiler_float32_equal(Int32_val(i), Int32_val(j));
+}
+
+value compiler_float32_of_float_boxed(value d)
+{
+    return caml_copy_int32(compiler_float32_of_float(Double_val(d)));
+}
+
+value compiler_float32_to_float_boxed(value i)
+{
+    return caml_copy_double(compiler_float32_to_float(Int32_val(i)));
+}
+
+static int compiler_float32_of_hex(const char * s, const char * end, float * res)
+{
+  /* See caml_float_of_hex */
```

Comment:
There is a fair amount of code duplication;
if it is not possible/desirable to factor out,
we should probably have a short comment
explaining why.

Reply:
Added a comment; we can't use a modified runtime in the flambda compiler build

Reply:
(The near-identical versions in float32.c are added in a later PR)

Reply:
(The near-identical versions in float32.c are added in a later PR)